pkgs/tools/typesetting/tex/texlive/combine.nix

```diff
@@ -241,6 +241,10 @@ in (buildEnv {
     then
       faketime $(date --utc -d@$SOURCE_DATE_EPOCH --iso-8601=seconds) luahbtex --output-directory="$TEXMFSYSVAR"/web2c/luahbtex -ini -jobname=lualatex -progname=lualatex lualatex.ini
     fi
+    if [[ -f "$TEXMFSYSVAR"/web2c/luatex/dvilualatex.fmt ]]
```

Comment:
Should we just stick `faketime` in front of `fmtutil`? I don't like that we are now rebuilding three formats that have just been built (sure, only 3 out of 59, but still). We have no guarantee that this list is or remains exhaustive anyway, so I say we go all in, and hope for the best.

Reply:
This list is should be exhaustive for *texlive.combine-scheme-basic*. For texlive.combined-scheme-full, the following paths also don't reproduce:
- /nix/store/8mnl94r0vcrlm7f41z92d5hlvl7akhry-texlive-combined-full-2022.20221227/share/texmf-var/web2c/tex/lollipop.fmt
- /nix/store/8mnl94r0vcrlm7f41z92d5hlvl7akhry-texlive-combined-full-2022.20221227/share/texmf-var/web2c/luahbtex/lualatex-dev.fmt
- /nix/store/8mnl94r0vcrlm7f41z92d5hlvl7akhry-texlive-combined-full-2022.20221227/share/texmf-var/web2c/luatex/dvilualatex-dev.fmt
- /nix/store/8mnl94r0vcrlm7f41z92d5hlvl7akhry-texlive-combined-full-2022.20221227/share/texmf-var/luatex-cache/context/7c7f8d725b1e9c75a3e798cb52b47225/trees/34e907e83e4734fdd0891aeea1ce848e.lua
- /nix/store/8mnl94r0vcrlm7f41z92d5hlvl7akhry-texlive-combined-full-2022.20221227/share/texmf-var/luatex-cache/context/7c7f8d725b1e9c75a3e798cb52b47225/trees/34e907e83e4734fdd0891aeea1ce848e.luc
- /nix/store/8mnl94r0vcrlm7f41z92d5hlvl7akhry-texlive-combined-full-2022.20221227/share/texmf-var/luatex-cache/context/7c7f8d725b1e9c75a3e798cb52b47225/trees/a3b666420e9dc4eaa86220c3fed48806.lua
- /nix/store/8mnl94r0vcrlm7f41z92d5hlvl7akhry-texlive-combined-full-2022.20221227/share/texmf-var/luatex-cache/context/7c7f8d725b1e9c75a3e798cb52b47225/trees/a3b666420e9dc4eaa86220c3fed48806.luc

The first three appear to reproduce when running fmtutil in “frozen time mode” (`fmtutil -f '1970-01-01 00:00:01'`), the current invocating only rewinds the clock to $SOURCE_DATE_EPOCH (1980-01-01 since there is no source?) and keeps it running. The remaining paths are produced by `mtxrun --generate` and also contain randomness in the form of uuids, so they cannot be made reproducible as easily.

Reply:
I don't think we necessarily need the frozen time mode – faketime says

> If the started process itself starts other (child) processes, they by default will start with the specified start-at-date again.

So when calling `faketime fmtutil`, each format is supposed to start from `SOURCE_DATE_EPOCH`. Could you please try @apfelkuchen6 and see if you get identical formats?

(I am worried about using the frozen clock. Bizarre bugs could slip through, unlikely as it may be, and we don't have the infrastructure to spot them. E.g. we shipped wrong hyphenation config for years, and nobody ever complained!)

Reply:
I tried with ` FORCE_SOURCE_DATE=1 faketime $(date --utc -d@$SOURCE_DATE_EPOCH --iso-8601=seconds) fmtutil --sys --all | grep '^fmtutil' # too verbose` and did not get identical formats:
```
✗ diff -r /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227 /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227.check
diff -r /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227/share/texmf-var/luatex-cache/context/337b2763f83e995fd8c4b2060cf8809b/trees/34e907e83e4734fdd0891aeea1ce848e.lua /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227.check/share/texmf-var/luatex-cache/context/337b2763f83e995fd8c4b2060cf8809b/trees/34e907e83e4734fdd0891aeea1ce848e.lua
236926c236926
<  ["time"]="13:34:31",
---
>  ["time"]="13:36:26",
236928c236928
<  ["uuid"]="9b453641-4be9-9170-7d05-7a6057638383",
---
>  ["uuid"]="8764dc9c-4457-b9d1-e341-bcee37e8ce30",
Binary files /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227/share/texmf-var/luatex-cache/context/337b2763f83e995fd8c4b2060cf8809b/trees/34e907e83e4734fdd0891aeea1ce848e.luc and /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227.check/share/texmf-var/luatex-cache/context/337b2763f83e995fd8c4b2060cf8809b/trees/34e907e83e4734fdd0891aeea1ce848e.luc differ
diff -r /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227/share/texmf-var/luatex-cache/context/337b2763f83e995fd8c4b2060cf8809b/trees/d1267bcc3565ade491b8cdefafea14b8.lua /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227.check/share/texmf-var/luatex-cache/context/337b2763f83e995fd8c4b2060cf8809b/trees/d1267bcc3565ade491b8cdefafea14b8.lua
140c140
<  ["time"]="13:34:31",
---
>  ["time"]="13:36:26",
142c142
<  ["uuid"]="e96defac-4ba0-ad1f-eb32-2d1d299e5bc9",
---
>  ["uuid"]="5bcb6981-4360-bd08-26c0-c7eb360a3a01",
Binary files /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227/share/texmf-var/luatex-cache/context/337b2763f83e995fd8c4b2060cf8809b/trees/d1267bcc3565ade491b8cdefafea14b8.luc and /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227.check/share/texmf-var/luatex-cache/context/337b2763f83e995fd8c4b2060cf8809b/trees/d1267bcc3565ade491b8cdefafea14b8.luc differ
Binary files /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227/share/texmf-var/web2c/luahbtex/lualatex.fmt and /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227.check/share/texmf-var/web2c/luahbtex/lualatex.fmt differ
Binary files /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227/share/texmf-var/web2c/luatex/dvilualatex-dev.fmt and /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227.check/share/texmf-var/web2c/luatex/dvilualatex-dev.fmt differ
Binary files /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227/share/texmf-var/web2c/luatex/dvilualatex.fmt and /nix/store/yhha457n8mdss0wfzaipb8qgf2pq14dl-texlive-combined-full-2022.20221227.check/share/texmf-var/web2c/luatex/dvilualatex.fmt differ
```
I also cannot reproduce the documented behaviour in faketime's manpage:
```
➜ SOURCE_DATE_EPOCH=12345                                                                                                                                 
➜ faketime $(date --utc -d@$SOURCE_DATE_EPOCH --iso-8601=seconds) sh -c 'date; sleep 10; date'
Thu Jan  1 04:25:45 AM CET 1970
Thu Jan  1 04:25:55 AM CET 1970
```

I tried recreating the formats in a loop:
```
    for fmt in tex/tex tex/lollipop {luahbtex/,luatex/dvi}lualatex{,-dev}; do
      if [[ -e "$TEXMFSYSVAR"/web2c/"$fmt".fmt ]]; then
        echo Recreating "$fmt".fmt...
        FORCE_SOURCE_DATE=1 faketime '1980-01-01 00:00:00' fmtutil --sys --byfmt ''${fmt##*/} | grep '^fmtutil'
      fi
    done;
```
But this also didn't result in reproducible results on my machine without `faketime -f` even though the same faketime invocation works when calling the engine directly (and not through fmtutil).

`SOURCE_DATE_EPOCH` in combine is indeed 315532800 -- 1980-01-01T00:00:00+00:00. Using the snapshot date is fine, but since hardcoding 1980 also works, I'm not sure why we wouldn't just use that (it allows for slightly more compact commands).

Reply:
Interesting. I think what we want is `TZ= faketime -f '@1980-01-01 00:00:00' fmtutil`:
```
[nix-shell:~]$ TZ= faketime -f '@1980-01-01 01:00:00' sh -c 'date; sleep 5; date'
Tue Jan  1 01:00:00 UTC 1980
Tue Jan  1 01:00:00 UTC 1980
```
whereas
```
[nix-shell:~]$ FAKETIME_DONT_RESET=1 TZ= faketime -f '@1980-01-01 01:00:00' sh -c 'date; sleep 5; date'
Tue Jan  1 01:00:00 UTC 1980
Tue Jan  1 01:00:05 UTC 1980
```
I understand `-f` is "advanced format" and `@` means "not frozen", so this should offer the best compromise: restart the clock of each process but not freeze it completely. We could hack it further with `faketime -f '@1980... x0.01` to slow down the clock. Just to be on the safe side.

Not sure if `TZ` is required, I would assume stdenv already sets it.

PS: it looks like the lua caches are trivial to patch, on the other hand. In a different PR though.